function/avro-import.py

```diff
@@ -0,0 +1,156 @@
+    error_file_name = subdirectory_name + "/error-" + start_time + ".txt"
+    write_to_bucket(error_file_name, message)
+    return message
+
+
+def write_upsert_to_bucket(job_id, content_string):
+    compressed_value = gzip.compress(bytes(content_string, 'utf-8'))
+    file_name = job_id + "/upsert.json"
+    write_to_bucket(file_name, compressed_value)
+
+
+def write_metadata_to_bucket(job_id, content_string):
+    file_name = job_id + "/metadata.json"
+    write_to_bucket(file_name, content_string)
+
+
+def write_to_bucket(file_name, content):
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket("avro-translated-json")
+    blob = bucket.blob(file_name)
+    blob.upload_from_string(content)
+
+
+def translate(reader, options=None):
+    if options is None:
+        options = {}
+    t = Translator(options)
+    return t.translate(reader)
+
+
+class Translator:
+    def __init__(self, options=None):
+        if options is None:
+            options = {}
+        defaults = {'b64-decode-enums': False, 'prefix-object-ids': False}
+        self.options = {**defaults, **options}
+
+    def translate(self, reader):
+        if reader is None:
+            return None
+
+        enums = _list_enums(reader.writer_schema)
+        results = [self._translate_record(record, enums)
+                   for record in reader if record['name'] != 'Metadata']
+        return results
+
+    def _translate_record(self, record, enums):
+        entity_type = record['name']
+        name = record['id']
+
+        def make_op(key, value):
+            if self.options['b64-decode-enums'] and (entity_type, key) in enums:
+                value = _b64_decode(value).decode("utf-8")
+            if self.options['prefix-object-ids'] and key == 'object_id':
+                value = 'drs://' + value
+            return _make_add_update_op(key, value)
+
+        attributes = [make_op(key, value)
+                      for key, value in record['object'].items() if value is not None]
+        relations = [make_op(relation['dst_name'],
+                             {'entityType': relation['dst_name'], 'entityName': relation['dst_id']})
+                     for relation in record['relations']]
+
+        return {
+            'name': name,
+            'entityType': entity_type,
+            'operations': [*attributes, *relations]
+        }
+
+
+def _b64_decode(encoded_value):
+    return base64.b64decode(encoded_value + "=" * (-len(encoded_value) % 4))
+
+
+def _list_enums(schema):
+    object_field = next(f for f in schema['fields'] if f['name'] == 'object')
+    types = [t for t in object_field['type'] if t['name'] != 'Metadata']
+    enums = {(entity_type['name'], field['name'])
+             for entity_type in types
+             for field in entity_type['fields']
+             for enum in field['type'] if isinstance(enum, dict) and enum['type'] == 'enum'}
+    return enums
+
+
+def _make_add_update_op(key, value):
```

Comment:
are you expecting to get list attributes? because this won't work for lists.

Reply:
@breilly2

Reply:
This is a good point. We haven't seen any yet, but I think we also may not have seen the entire range of shapes of data we might possibly get. This has been a point of discussion with the Gen3 people. Definitely something to keep an eye on.

Reply:
In general, you can't get lists in through the TSV import anyway -- with the very large exception of populating set types. There are certainly other use cases that are hitting the Rawls endpoint to populate with lists; I get personally contacted by a DSP-internal Broadie about it every 6 months or so. It stands to reason that plenty more people out there are doing it too, given I'm one developer getting my intel from unofficial channels.